oeps/oep-0045-arch-ops-and-config.rst

```diff
@@ -0,0 +1,175 @@
+
+
+.. _native Dockerfile syntax: https://docs.docker.com/engine/reference/builder/#dockerfile-reference
+
+**Docker Images**
+
+edX will provide Docker images for applications that captures the latest code on the master branch as well as images representing named releases. edX will not provide these images for named releases prior to the acceptance and implementation of this OEP which is Juniper at time of writing.
+
+
+Configuration
+*************
+ 
+**Django settings**
+
+Having a single artifact that runs with different configurations increases stability by improving development parity with other deployment environments. edX applications already support configuration overrides via a yaml file for production environments, but development and test environments tend to configure the application using different code paths via a settings/devstack.py or settings/test.py file.
+
+Additionally it is not clear which settings are required to be overridden and which settings have values that may technically work but are inappropriate for production systems. To alleviate these issues edX django applications will adopt the following settings structure:
+
+.. code-block:: text
+
+  settings
+  ├── __init__.py
+  ├── required.py
+  └── defaults.py
+
+
+*  ``__init__.py`` - Sourcing our config from this file within the settings directory takes advantage of django defaults and means that settings will be picked up automatically without needing to specify ``--settings`` anywhere. This entry point would import ``required.py``, ``defaults.py``, and the code to override both from a config file.
+*  ``required.py`` - all settings which are required to run and do not have a reasonable production-ready default, e.g. LMS_BASE_URL which will be different per environment.
+*  ``defaults.py`` - other settings which will have production-ready defaults
+
+The settings defined in ``required.py`` and ``defaults.py`` files are mutually exclusive, representing all application specific settings as well as installed library settings whose values either must be provided or whose defaults are not considered production-ready.
+
+``required.py`` variables must be overridden by operators.  The application will check that operators provided these values, and will not start unless they are set. This allows operators to fail fast rather than finding out about an unset value when users exercise those breaking codepaths. Application developers are encouraged to keep the list of required settings to a minimum.
+
+This new settings structure obviates the need for any other python files in the settings directory (such as ``devstack.py``, ``test.py``, etc). The values currently set in those files should be moved to a corresponding ``devstack.yml``, ``test.yml``, etc in the same settings directory. This gives developers and operators more consistency across environments since the same code paths are being executed with different values.
```

Comment:
Developers have found "private.py" helpful.  That will also go away?

Reply:
Looking at edx-platform/lms/envs/test.py, there is a little bit of executable code in it, for example to randomize keys.  How will that be handled in a test.yml world?

Reply:
Yes.  However we should talk about how to replicate it's usefulness in a `yml` config world.

Reply:
FWIW, I took a look back at how we handled this generally (custom logic in settings `.py` files) on the Stanford fork...

We used to have some crazier things (heck, we shadowed the entire `{lms/cms}/envs/` directory structure), but here's what we were actually using in practice, at the end:
- Contextually removing some settings (`INSTALLED_APPS`, `MIDDLEWARE_CLASSES`) [1]
- expanding wildcard paths [2]
- overriding logging config [3]

Re: `private.py` specifically:
We definitely used this and I'd worry that folks won't be able to replicate all of the capabilities of custom Python logic w/in a `.yml` file. Is there a specific reason we're looking to remove this feature, eg, "`private.py` considered harmful"?
Or is this more "getting caught up in the storm" of cleaning out the settings directory?
Since `private.py` is explicitly `git-ignored`, it shouldn't increase the number of files checked in to that directory, with minimal maintenance overhead on the implementation [4].

I'm open to alternatives, as long as we don't have to reinvent the wheel.

- [1] https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/cms/envs/test.py#L8-L11
- [2] https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/cms/envs/common.py#L67-L69
- [3] https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/lms/envs/devstack.py#L9-L20
- [4] https://github.com/Stanford-Online/edx-platform/blob/master/lms/envs/devstack.py#L273-L275

Reply:
@nedbat I just looked through that file and I didn't see any executable code that couldn't just be a preset value.  I don't think we will be able to get away with just a straight import and override from the yaml, but I do think that we can keep any executable code required to set up the app in one code path.

If people have valid cases for executable code in the settings files that only apply to a particular environment I would love to see those cases.  I wasn't able to find any myself.

Reply:
@stvstnfrd thanks for all those great references!  I will copy a line written in previous comment

> One of the goals of this OEP is to have an application that executes the same code paths in development as in production wherever reasonable. I believe that this is a place that could reasonably be one code path.

I would love for that to be our starting assumption and see if there is a way to achieve the needs of operators and developers without compromising current functionality (though we may need to change current processes/mindsets).  Your examples bring up a great and very common configuration challenge.  `INSTALLED_APPS` and `MIDDLEWARE_CLASSES`

Would it be reasonable for all configuration files to be able to define:
```
INSTALLED_APPS:
- 'app1'
- 'app2'
APPEND_INSTALLED_APPS:
- 'app3'
REMOVE_INSTALLED_APPS:
- 'app2'
```


`INSTALLED_APPS` would completely overwrite the value.  `APPEND` and `REMOVE` would take the corresponding list of value and perform that action.  Allowing the code path to resolve these common cases the same way across the board.

Reply:
@bderusha Cool, yeah, I probably should have split apart the two lines of thought in that comment (`private.py` vs general config), as I am implying the former could be be kept around w/ negligible overhead/maintenance, but not implying as much with the latter.

For the general config stuff, I didn't mean to imply "this can't be done!"; just wanted to flag some examples that may not yet be accounted for :)

Reply:
@bderusha Your `INSTALLED_APPS` example does a good job of highlighting how handling some Django settings will be more complex than just "fully override it with the value in the YAML". I think [edx-platform OPTIONAL_APPS](https://github.com/edx/edx-platform/blob/master/lms/envs/common.py#L3152-L3196) also illustrates this.

I think it would be worth eventually specifying our "special" Django settings (`APPEND_INSTALLED_APPS`, `REMOVE_MIDDLEWARE`, etc.) somewhere visible, with a reference implementation of how to handle them in `__init__.py`. That being said, I think it's out of the scope of this PR, as it would require some digging and discovery to find out what those special settings should be and how we would handle them. Maybe that's something that would end up in your follow-up PR.